src/python/grpcio/grpc/experimental/aio/_channel.py

```diff
@@ -13,15 +13,207 @@
+    `UnaryUnaryMultiCallable` object is called.
+    """
+
+    def __init__(self, aio_call: asyncio.Task,
+                 response_deserializer: DeserializingFunction,
+                 aio_call_cancel_status: cygrpc.AioCallCancelStatus):
+        self._exception = None
+        self._response = None
+        self._code = None
+        self._details = None
+        self._initial_metadata = None
+        self._trailing_metadata = None
+        self._cancelled = False
+        self._aio_call = aio_call
+        self._aio_call_cancel_status = aio_call_cancel_status
+        self._response_deserializer = response_deserializer
+
+    def __del__(self):
+        self.cancel()
+
+    def cancel(self) -> bool:
+        """
+        Cancels the ongoing RPC request.
+
+        Returns:
+          True if the RPC can be canceled, False if was already cancelled or terminated.
+        """
+        if self.cancelled() or self.done():
+            return False
+
+        details = 'Locally cancelled by application!'
+        code = grpc.StatusCode.CANCELLED
+        self._aio_call_cancel_status.cancel(
+            _common.STATUS_CODE_TO_CYGRPC_STATUS_CODE[code], details=details)
+        self._aio_call.cancel()
+        self._details = details
+        self._code = code
+        self._cancelled = True
+        return True
+
+    def cancelled(self) -> bool:
+        """
+        Returns if the RPC was cancelled.
+
+        Returns:
+          True if the requests was cancelled, False if not.
+        """
+        return self._cancelled
+
+    def running(self) -> bool:
+        """
+        Returns if the RPC is running.
+
+        Returns:
+          True if the requests is running, False if it already terminated.
+        """
+        return self._code is None
+
+    def done(self) -> bool:
+        """
+        Returns if the RPC has finished.
+
+        Returns:
+          True if the requests has finished, False is if still ongoing.
+        """
+        return self._code is not None
+
+    async def initial_metadata(self):
+        raise NotImplementedError
+
+    async def trailing_metadata(self):
+        raise NotImplementedError
+
+    async def code(self) -> grpc.StatusCode:
+        """
+        Returns the `grpc.StatusCode` if the RPC is finished, otherwise first waits till the
+        RPC finishes.
+
+        Returns:
+          The `grpc.StatusCode` status code.
+        """
+        if not self.done():
+            try:
+                await self
+            except (asyncio.CancelledError, AioRpcError):
+                pass
+
+        return self._code
+
+    async def details(self) -> str:
+        """
+        Returns the details if the RPC is finished, otherwise first waits till the
+        RPC finishes.
+
+        Returns:
+          The details.
+        """
+        if not self.done():
+            try:
+                await self
+            except (asyncio.CancelledError, AioRpcError):
```

Comment:
I would recommend add a comment here, saying that if RPC throws exception, it is expected that the `code`, `details` are filled by the thrown exception.

Also, we should make default value to `UNKNOWN`, check if `code` and `details` are set before returning.

Reply:
I would add the UNKNOWN as a default value, not sure what you mean by

> check if code and details are set before returning.

In any case let's double-check once the new commit is added to this PR with support for RPC state